### ISSUE TITLE
Order of statements using Execute.WithConnection() and Execute.Sql()

### ISSUE THREAD

**Issue by igitur** [fluentmigrator/fluentmigrator#1332]
**Describe the question**
I'm trying to use `.patch` files to migrate my MSSQL stored procedures instead of describing the full stored procedure for UP and DOWN migrations. (Side note: because I'm simultaneously dealing with many code branches and using the traditional approach I found that after merging branches, some stored proc migrations might undo older changes. And patch files can be applied in reverse too, which simplifies the DOWN migration).

Anyway, to do this I have to extract the current script from MSSQL (via `sp_helptext`)

So, in short, I have this migration:
```c#
public override void Up()
{
    string newScript = "";
    Execute.WithConnection((connection, tx) => newScript = GetStoredProcTextAndApplyPatch());

    if (string.IsNullOrWhiteSpace(newScript))
        Console.WriteLine("No script to execute");
    else
    {
        Console.WriteLine("Executing change script");
        Execute.Sql(newScript);
    }
}
```

When migrating (using console runner), it seems the bottom block executes first, hence `newScript` is empty and the `Execute.WithConnection` executes after that. Some kind of delayed lambda function?

I guess I can execute the new script inside the `.WithConnection` block, but then I don't have access to the fluent features.

Is this behaviour intended?

**Documentation Pages You've Read So Far**
I browsed through the documentation and open issues on Github

**Expected benefit**
Statements to be executed in order in which they appear in the code.

**Information (please complete the following information):**
 - OS: Windows 10
 - Platform: .NET 4.7.2
 - FluentMigrator version [ e.g 3.2.9 ]
 - FluentMigrator runner [FluentMigrator.Console 3.2.9]
 - Database Management System [ SQLServer ]
 - Database Management System Version [ SQLServer 2016 ]

**Additional context**

**Comment by igitur** [comment]
BTW, I couldn't find any unit tests for `.WithConnection()`. Did I just not search hard enough?

**Comment by jzabroski** [comment]
That's a really interesting approach to versioning stored procedures.  

Execute.WithConnection does execute _its' contents_ "out of sequence", in a way.  The issue is that FluentMigrator collects all "commands" first, then executes them.  So, FluentMigrator merely collected _the promise_ to execute some lambda, not the _result_ itself.  This can lead to some counter-intuitive behavior when mixing DDL statements with Execute.WithConnection.  The important thing to realize is `Up()` merely collects a series of promises.  When you read `newScript`, it is no longer in the same scope, because it gets closed over and will get saved until the lambda actually executes.

I had loosely proposed a monadic interface for FluentMigrator, but its a big project and refactoring, so I never got to working on it.  It was actually a big reason why I got involved in the project in the first place, because I was moving code from FluentMigrator 1.x to 3.y, and in 1.x, the Transaction item was allowed to be null. See: https://github.com/fluentmigrator/fluentmigrator/issues/987 In particular, the following comment explains how I would like to capture these "side effects": https://github.com/fluentmigrator/fluentmigrator/issues/987#issuecomment-468810759

**Comment by jzabroski** [comment]
BTW, the workaround is to simply move your Execute.WithConnection into its own migration.  Its not a 100% workaround because you can't capture the side-effects of the computation itself.  The other workaround is to not directly couple your connection logic to IMigration interface, and instead write delegates that take a IDbConnection and IDbTransaction as parameters, because then you can daisy chain your logic _within the Execute.WithConnection_ code.

**Comment by jzabroski** [comment]
I suppose if we implemented #854 , we could each top-level command return a Task, and have a way to ContinueWith the migration.  Perhaps then a `With.Task` or something could be used to get any results from Execute.WithConnection.  This would be more possible with return type covariance mentioned for C# 9.0 https://devblogs.microsoft.com/dotnet/welcome-to-c-9-0/

However, it would also mean dropping support for legacy .NET Framework, etc.  So it's realistically not going to be a change I'll want to make for at least two years.

**Comment by jzabroski** [comment]
@igitur Curious how you made out?

**Comment by igitur** [comment]
Promise I'll give feedback when I do anything. Have been working on other stuff in the meantaime. Thanks for the replies though.

**Comment by igitur** [comment]
> BTW, the workaround is to simply move your Execute.WithConnection into its own migration.

Unfortunately this isn't a viable solution to me, because the 2 "parts" form an atomic operation. Even if I were to do it with the up migration, it would break the down migration.

My 2 "parts" are:
- Get stored procedure script using `sp_helptext` and patch it. This all runs in `Execute.WithConnection`
- Run `Execute.Sql()` with the patched script.

So in the down migration case, it should be the same steps in the same order. It's just the application of `patch` that reverses. (`-r` parameter for standard GNU `patch` command)

In the end I just solved this by not relying on `Execute.Sql()` and handling the entire migration inside a single `Execute.WithConnection` block, using Dapper to execute the patched SQL script.

This patched-based approach is working really well. I'm using it for complex views and stored procedures. It could theoretically work for UDFs too, but I haven't tried. Maybe I'll release it as a FluentMigrator extension package.

**Comment by jzabroski** [comment]
> Maybe I'll release it as a FluentMigrator extension package.

Sure.  Update the README with a link to third party projects if/when you do.

I personally don't think patch-based approach should go into core, but I do want some built-in support for views and stored procedures.  The problem is that you would then need to configure the "layout" for the text in those files.  What I do in my own projects is:

* FluentMigrations.csproj
  * /Schema folder
    * /Functions folder
    * /Security folder
      * db_datawriter.sql (contains a script to grant all permissions associated with this role - runs in an AfterAll Maintenance Profile)
      * db_datareader.sql (ibid)
    * /Stored Procedures folder
      * schemaName.ObjectName.sql (uses CREATE OR ALTER for newer versions of SQL Server that support this, as this maintains created on / last modified audit log in INFORMATION_SCHEMA.ROUTINES)

Hope that helps.

**Comment by jzabroski** [comment]
> Unfortunately this isn't a viable solution to me, because the 2 "parts" form an atomic operation. Even if I were to do it with the up migration, it would break the down migration.

Yes, this is what I meant in "the other workaround" (re-quoting below):


> **The other workaround** is to not directly couple your connection logic to IMigration interface, and instead write delegates that take a IDbConnection and IDbTransaction as parameters, because then you can daisy chain your logic _within the Execute.WithConnection_ code.

I actually do this workaround more these days than not, as I don't want my logic to be directly coupled to FluentMigrator, per say.  For example, I have functions that take an IDbConnection and IDbTransaction and refresh a database object in a SQL Server database by calling [`sp_refreshsqlmodule`](https://docs.microsoft.com/en-us/sql/relational-databases/system-stored-procedures/sp-refreshsqlmodule-transact-sql?view=sql-server-ver15) and I have a separate function that takes an IDbConnection and IDbTransaction that gets the transitive closure of all objects to refresh.  My Execute.WithConnection then calls the function to get the transitive closure, and then walks that transitive closure and calls the function that calls sp_refreshsqlmodule.

**Comment by jzabroski** [comment]
BTW, this may seem dumb, but - rather than use patches, why not just dump a copy of the entire object tree into a folder structure where the parent folder is a release name.  Then your migration simply computes a sha1 hash sum on the most recent release name and the previous release name, and does a file comparison for differences.  Then you avoid the complexity of figuring out how a patch got applied, and potentially any bugs with a patch not working since you're doing a live merge operation in a production app (which is why I think your approach is interesting).
